### PR TITLE
Convert `NaN` to `None` when writing JSON

### PR DESCRIPTION
**Pull request by rusheb-apollo**
## Description
- Convert `nan` values to `None` before writing treetrace files

## Related Issue
[ape #439](https://github.com/ApolloResearch/ape/issues/348) 

## Motivation and Context
This fixes an issue which caused the treetrace viewer to crash when the tracing output contained `nan` values. The most common instance of this was when `eval.run` would return a dict where one of the values was `np.nan` due to trying to take the mean of an empty list. 

### Cause of the problem
- Treetrace writes its tree to a JSON file. When [writing the `result` field](https://github.com/ApolloResearch/interlab/blob/rs-avoid-nan/treetrace/tracing/tracingnode.py#L206-L209), json_dumps has `allow_nan=True` (the default)
- This means that `nan` values end up in the JSON file
- Somewhere downstream of this, the JSON has been loaded back to a dict and the starlette package tries to [serialize it again, but this time with `allow_nan=False`](https://github.com/encode/starlette/blob/5a1bec33f8d6a669a3670f51034de83292d19408/starlette/responses.py#L186-L193). This is what causes the treetrace viewer to barf. 

- So this solution just converts `nan` values to `None` instead to avoid this problem. 

### Alternatives considered
- Catch the exception from starlette and skip the record -- I didn't do this because it seemed harder and also might lead to confusingly missing rows in the treetrace output
- Prevent `run` from returning `nan` or convert it from within `ape.run` or somewhere -- this seems wrong because I think it's the responsibility of TreeTrace to write valid JSON for itself to read later 

## How Has This Been Tested?
Manually: I pointed my local pyproject.toml at this branch and ran a sample which was previously causing the error

## Does this PR introduce a breaking change?
No

### CODE REVIEW

**Comment by mikita-apollo** [comment]
Nice, do we expect this to fix [Alex's issues](https://apolloresearchhq.slack.com/archives/C068QH5KXGA/p1718189180797679)?

File: treetrace/tracing/storage.py [hunk 1]
```diff
@@ -140,8 +157,10 @@ def _write_node_into(self, directory: str, node: TracingNode, root_node: bool):
         if node.directory:
             self._write_node_dir(directory, node, root_node)
         else:
-            data = json.dumps(node.to_dict(root=root_node)).encode()
-            data_root = json.dumps(node.to_dict(False, root=root_node)).encode()
+            data = json.dumps(node.to_dict(root=root_node), cls=NanConverter).encode()
+            data_root = json.dumps(
+                node.to_dict(False, root=root_node), cls=NanConverter
```
Comment (mikita-apollo) [inline treetrace/tracing/storage.py:162]: Any chance we can use a keyword argument for this `False`? 

Hard to figure out what it stands for when doing a review

**Review by mikita-apollo — APPROVED** [review]
LGTM, one comment is up to you

File: treetrace/tracing/storage.py (hunk 1 above)
Comment (rusheb-apollo) [reply treetrace/tracing/storage.py:162]: I don't think it should be configurable, but I'll add a descriptive variable.

Comment (rusheb-apollo) [reply treetrace/tracing/storage.py:162]: Sorry, misunderstood what you meant in my previous response. Happy to do this, thx

**Pull request merged**